### PR TITLE
Io excel

### PR DESCRIPTION
**Pull request by bashtage**
- [ ] Tests added: Please use `assert_type()` to assert the type of any return value

### CODE REVIEW

**Comment by bashtage** [comment]
No new tests yet. Existing pass except for one buggy one, which was fixed.

File: pandas-stubs/io/excel/_base.pyi [hunk 1]
```diff
@@ -1,188 +1,247 @@
...
     index_col: int | Sequence[int] | None = ...,
-    usecols: int | str | Sequence[int | str | Callable] | None = ...,
-    squeeze: bool = ...,
-    dtype: str | dict[str, Any] | Dtype = ...,
-    engine: str | None = ...,
-    converters: dict[int | str, Callable] | None = ...,
-    true_values: Sequence[Scalar] | None = ...,
-    false_values: Sequence[Scalar] | None = ...,
-    skiprows: Sequence[int] | int | Callable | None = ...,
+    usecols: str | Sequence[int] | Sequence[str] | Callable[[str], bool] | None = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/io/excel/_base.pyi:44]: `Sequence[str]` will match `str` so no need to include both.

File: pandas-stubs/io/excel/_base.pyi [hunk 2]
```diff
@@ -1,188 +1,247 @@
...
+    dtype: str | Dtype | dict[str, str | Dtype] | None = ...,
+    engine: Literal["xlrd", "openpyxl", "odf", "pyxlsb"] | None = ...,
+    converters: dict[int | str, Callable[[object], object]] | None = ...,
+    true_values: Iterable[Hashable] | None = ...,
+    false_values: Iterable[Hashable] | None = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/io/excel/_base.pyi:49]: I think `Iterable[Hashable]` is too wide.  Docs say "list", so I think `list[str | int]` would be sufficient.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]

**Comment by Dr-Irv** [comment]
> Is it OK to include optional dependencies in the typing stubs? Will this cause problems downstream?

The dependencies you added are there just for testing.  If there are typing stubs for the optional dependencies, then I think we should include them as required dependencies for pandas-stubs.

File: pandas-stubs/io/excel/_base.pyi (hunk 2 above)
Comment (bashtage) [reply pandas-stubs/io/excel/_base.pyi:49]: The iterable comes from pandas main. I also tested using an iterator and it worked as expected.

**Review by Dr-Irv — COMMENTED** [review]
I'm fine with all the changes, but wondering if we should also move the `test_read_excel()` function in `test_frame.py` to `test_io.py` .  Could do this as part of this PR or as a follow-up PR.

**Comment by Dr-Irv** [comment]
Also need to resolve conflicts

File: pandas-stubs/io/excel/_base.pyi (hunk 1 above)
Comment (bashtage) [reply pandas-stubs/io/excel/_base.pyi]: Fixed.

**Review by Dr-Irv — APPROVED** [review]
thanks @bashtage

**Comment by Dr-Irv** [comment]
@bashtage can you resolve conflicts, then ping when green.

**Comment by bashtage** [comment]
Green, ignoring the pre -test fail

**Comment by Dr-Irv** [comment]
UGH.  I tried to resolve the conflicts via the web editor, and messed up black formatting.  Can you pull what I did and fix it?

**Review by Dr-Irv — APPROVED** [review]
thanks @bashtage

**Pull request merged**